src/scripts/profiling/run_profiling.py

```diff
@@ -22,14 +26,56 @@ def current_time(formatstr: str = "%Y-%m-%d_%H%M") -> str:
     return datetime.utcnow().strftime(formatstr)
 
 
+def record_run_statistics(output_file: str, s: Simulation) -> None:
+    """
+    After concluding the profiling session, but before cleanup,
+    record statistics from the simulation outputs that we wish
+    to present alongside the profiler outputs.
+
+    Key / value pairs are:
+    pop_df_size: (rows, cols)
+        Number of rows and columns in the final population dataframe
+    pop_df_mem_mb: MBs
+        Size in MBs of the final population dataframe
+    pop_df_times_extended: n
+        Number of times the population dataframe had to be expanded
+    """
+    # Initialise empty dict to add statistics to,
+    # then save in JSON format
+    stats_to_record = dict()
+
+    # Record statistics as [key, value] pairs
+
+    # Population dataframe statistics
+    pops = s.population
+    stats_to_record["pop_df_rows"] = pops.props.shape[0]
+    stats_to_record["pop_df_cols"] = pops.props.shape[1]
+    stats_to_record["pop_df_mem_mb"] = (
+        pops.props.memory_usage(index=True, deep=True).sum() / 1e6
+    )
+    stats_to_record["pop_df_times_extended"] = int(
+        np.ceil((pops.props.shape[0] - pops.initial_size) / pops.new_rows.shape[0])
+    )
```

Comment:
Minor nitpick but I would say using a dict literal here is a bit cleaner as avoids repetition of `stats_to_record` variable name and [should be a bit more performant](https://gist.github.com/hofrob/ad143aaa84c096f42489c2520a3875f9) (not that performance here is particularly important)

```suggestion
    # Population dataframe statistics
    pops = s.population
    stats_to_record = {
        "pop_df_rows": pops.props.shape[0],
        "pop_df_cols":  pops.props.shape[1],
        "pop_df_mem_mb": (
            pops.props.memory_usage(index=True, deep=True).sum() / 1e6
        ),
        "pop_df_times_extended": int(
            np.ceil((pops.props.shape[0] - pops.initial_size) / pops.new_rows.shape[0])
        )
    }
```